### DISCUSSION TITLE
Specific use case: how to contact to a BirdNET-Pi in the field without static IP adress (LTE router)

### DISCUSSION THREAD

**Discussion by DD4WH** [mcguirepr89/BirdNET-Pi#56] [Q&A] [answered]
UPDATE: THIS SOLUTION IS OUTDATED AND DOES NOT WORK ANY MORE WITH NEWER VERSIONS OF BIRDNETPI
USE THIS SOLUTION INSTEAD:
https://github.com/mcguirepr89/BirdNET-Pi/discussions/161#discussion-3899995

+++++++++++++++++++++++++++++++++++++++++++++
I have one specific use case for the BirdNET-Pi in an ornithologically interesting area. I have a little hut there with electricity, but NO internet connection. How can I connect to this BirdNET-Pi from my home network?

![grafik](https://user-images.githubusercontent.com/14326464/138687284-686c59d5-b22e-4c17-b147-c95a36921954.png)

One solution that I found (which I tested and realized that it works, but which is not yet perfect). Note that this is a very special use case and a very special solution (I use a FritzBox as my home router, where I am able to obtain a dynDNS service):

* LTE-router: I bought a cheap LTE router and bought a SIM card with a cheap data rate (1GB per month). This router can connect to the internet by 4G/LTE and is connected to the RPi by a LAN cable. The Pi is thus connected to the internet.
* However, the router does not have a permanent static IP adress, so it was not possible for me to connect to the Pi via NoMachine or VNC (with Anydesk it would be possible, but Anydesk does not run on a 64bit RPi OS)
* Solution: I connect the RPi via VPN client with my home router, which serves as a VPN server (FritzBox) --> when the Pi is connected to my home router via VPN, I can reach it within my home network

1.) Install vpnc on the RPi: `sudo apt-get install vpnc`
2.) Configure the conf-file: `sudo nano /etc/vpnc/fritzbox.conf`

`IPSec gateway xxxxxxx.myfritz.net`
`IPSec ID your_VPN_name`
`IPSec secret xxxxxxxxxxxxxxxxxxxx`
`IKE Authmode psk`
`Xauth username xxxxxx`
`Xauth password xxxxxxxxxxxxxxxxxxxx`
`local port 0`
`DPD idle timeout (our side) 0`

3.) Connect the VPN: `sudo vpnc fritzbox.conf`
4.) Disconnect the VPN: `sudo vpnc-disconnect`
5.) enable VPN on boot: `sudo nano /etc/rc.local`
Insert these two lines:

`sleep 10 # wait until network is ready!`
`vpnc fritzbox.conf`

6.) Reconnect network & VPN, if network connection is lost (which occurs often with RPi, especially with WiFi)

`sudo nano /usr/local/bin/checkwifi.sh`

`#!/bin/bash`
`ping -c4 MY.ROUTERS.IP.ADRESS > /dev/null` # put your routers IP adress here
 `if [ $? != 0 ]` 
`then`
`  sudo /sbin/shutdown -r now`
`fi`
`exit`

Then add this line into your crontab file to check the connection every 5 minutes

`crontab -e`

`#DD4WH / check internet connection and reboot if necessary`
`*/5 * * * * /usr/bin/sudo -H /usr/local/bin/checkwifi.sh >> /dev/null 2>&1`

I am not sure, whether you need to assign rights to the shell file!?
However, I did this and it at least did not hurt:

`chmod 4755 /usr/local/bin/checkwifi.sh`

It now checks every 5 minutes whether the router in the home network can be pinged. if not, it reboots the RPi. While rebooting, the internet connection is established and the vpnc VPN service is restarted.

With the help by Patrick (see below) this seems to work now!!! Thanks a lot!

I can now connect to the RPi inside my home network by NoMachine and look/listen to the browser output.
Will test it under real field conditions in the next weeks.

UPDATED 2021/10/25 to incorporate Patricks hints

I used code from the following resources:

https://www.kuemmel.wtf/?p=363
https://forum-raspberrypi.de/forum/thread/48999-vpnc-im-autostart/
https://weworkweplay.com/play/rebooting-the-raspberry-pi-when-it-loses-wireless-connection-wifi/

**Comment by mcguirepr89** [comment]
I'll elaborate when I'm home, but:

```
until ping -c 1 8.8.8.8;do
sleep 3
done
```
Then whatever command you need


This will wait for internet before proceeding
`ping` by itself fails in scripts without the "-c" count flag to limit its attempts

**Reply by DD4WH** [reply]
I used:

`ping -c4 192.168.1.1 > /dev/null`
 
`if [ $? != 0 ] `
`then`
`  sudo /sbin/shutdown -r now`
`fi`

But ping does not execute, when in a shell file.
The very same command does execute in the terminal.

**Comment by DD4WH** [comment]
It seems my method described above does not work any more with the latest version of BirdNET-Pi. I do not know why. There is a different Crontab file now? And on every command in the terminal I get "could not resolve hostname birdnetpi"?
Will try again with a clean install next week.

**Reply by mcguirepr89** [reply]
The resolving hostname error has to do with the VNC installation _seeming_ to go fine, but likely didn't set the hostname properly. That makes the most sense to me.

**Reply by DD4WH** [reply]
Oh, that sounds like a good explanation, I will check that! Thanks!

**Comment by mcguirepr89** [comment]
What shebang are you using for the shell script?

**Comment by mcguirepr89** [comment]
I am finally home. It seems you may have given up on this, but if not:

placing "set -x" at the top of your shell script just below the shebang will enable shell debuging:
```
#!/usr/bin/env bash
set -x
ping -c4 8.8.8.8
```
`ping 8.8.8.8` (Google's DNS) will check for internet connectivity, while `ping 192.168.1.1` (if my assumptions about where that IP address came from) will only check for a connection to the router.

```
cat << EOF > /home/pi/testingping.sh
#!/usr/bin/env bash
set -x
ping -c4 192.168.1.1
EOF
```
then
```
chmod +x /home/pi/testingping.sh
```
then
```
/home/pi/testingping.sh
```
That should show you _why_ the ping seems to fail.

**Comment by DD4WH** [comment]
Patrick, thanks a lot for your hints! Will try to go through them systematically later this week. It was too late yesterday to continue debugging.

Yes, my idea was to check whether the internet connection is there AND the VPN connection works properly. So I thought it would be good to check connection to the IP adress of my home router (which can only be reached if the internet connection AND VPN tunnel work).

I had to look up, what a "shebang" is, have not yet fully understood its purpose. If I store a file as  `/usr/local/bin/checkwifi.sh`, does it have to have a shebang like this at the top of the file ?

`#! /usr/local/bin/checkwifi.sh`

Sorry for my questions, which seem to be on very basic Linux topics, I fear. I am learning a lot from all your help!

**Comment by DD4WH** [comment, marked as answer]
@mcguirepr89 : Patrick, thanks a lot for your awesome hints!

After a good night of sleep everything turned out well and now fully working as expected!

What I did wrong and had to correct:
* typo in router ID
* I had to add `#!/bin/bash` in the first line of `/usr/local/bin/checkwifi.sh`
* After that, checking connection to my home routers´ IP worked --> if the RPi looses connection to my home network (which is checked every 5minutes), it reboots the RPi, leading to a reconnection of the LAN connection and restarts the vpnc VPN service 

I will now put in this additional info into the first post in this thread.

**Reply by mcguirepr89** [reply]
I'm really glad I could be of help. I was thinking about this more, and I might have another solution. 

The LTE router definitely has a public IP address (that is how it is reached and how you know it's online) -- the issue is that it is dynamic, likely new every power cycle. The possible solution I think may work would be setting up a postfix mailserver (standard baked in linux stuff that just needs configuration) that can email you with the LTE router's public IP address daily/hourly/whatever is needed. Then you should be able to use NoMachine to access the desktop remotely, or alternately, I could probably cook up something that will adjust your Caddyfile when that IP changes to make the installation reachable at whatever that IP address turns out to be.

To discover the LTE public IP address at any given time, _from the Pi_ issue `curl -4 ifconfig.co`. The results will be the public IP address for the router.

Let me know if you'd like to explore this possible solution any further.

My best!

PS -- please take a look at the annotation for the "Headless VNC Installation" if you have a chance and let me know if there is anything missing or unclear.

**Comment by mcguirepr89** [comment]
This is the most elegant solution to the dynamic dns problem -- and it provides a free domain name for all users! 
### [NoIP.com](https://www.noip.com/)
I will provide screenshot step-by-step guides for this, but will ultimately tuck this into an optional service (since it's just an executable and only supports init and not systemd).
1. Registered a FREE hostname using their FREE provided domain 
1. Installed the Dynamic Update Client onto my BirdNET-Pi

That's it!! Now, the Dynamic Update Client (on the BirdNET-Pi) will update the DDNS (NoIP) that it has a new IP address, and the A record (the public IP address associated with the free No-IP hostname). This is basically what AnyDesk does to keep its clients updated with its servers.

I am mocking up a diagram to explain how it works: I will update this shortly and make a discussion for all users who may suffer from this issue and/or want to use this solution as a free way to share their installations who may have dynamic public IP addresses (@Christoph-Lauer)

Diagram:
```
~~ How to manage a Dynamic Public IP Address ~~
IP=Public IP Address (find with `curl -4 ifconfig.co`)
DDNS=Dynamic Domain Name Server (https://noip.com)
DUC=Dynamic Update Client (on the BirdNET-Pi)
(This diagram does not show network traffic flow, 
rather, it shows how the dynamic dns is handled)

+-----------------------------------+
|                                   | 
|    Dynamic Domain Name Server     |  
|      No-IP hostname entry for     |
|        birdnetpi.ddns.net:        |
|"A" record target ->172.73.170.186 |      
|                                   |       +---------------+
+-----------------------------------+<----->|    Internet   |
                 ^|                         |ISP/LTE Carrier|
                 || DDNS directs all        +---------------+
DUC updates DDNS || requests to                     |
when IP changes  || birdnetpi.ddns.net              | ISP or LTE carrier
                 || to 172.73.170.186               | issues DHCP IP to
                 |v                                 | modem or router
+-----------------------------------+               v
|                                   |       +----------------+
|            BirdNET-Pi             |       |  Router/Modem  |      
|      Public IP=172.73.170.186     |<----->|Passes its IP to|
| [ Running DUC ] checks every Nmin |       |the rest of the |
|    Hostname=birdnetpi.ddns.net    |       |     network    |
|                                   |       +----------------+
+-----------------------------------+
```
When the ISP/LTE Carrier issues a new public IP address to the Modem/Router, the DUC running on the BirdNET-Pi sends an update to the DDNS that it has a new IP address. The DDNS updates its A record so that requests to "http://birdnetpi.ddns.net" resolve to the newly updated IP address.
